config/stack.go

```diff
@@ -40,12 +47,251 @@ type Unit struct {
 	Values *cty.Value `hcl:"values,attr"`
 }
 
+// Stack represents the stack block in the configuration.
+type Stack struct {
+	Name   string `hcl:",label"`
+	Source string `hcl:"source,attr"`
+	Path   string `hcl:"path,attr"`
+}
+
+// GenerateStacks generates the stack files.
+func GenerateStacks(ctx context.Context, opts *options.TerragruntOptions) error {
+	processedFiles := make(map[string]bool)
+	// initial files setting as stack file
+	foundFiles := []string{opts.TerragruntStackConfigPath}
+
+	for {
+		// check if we have already processed the files
+		processedNewFiles := false
+
+		for _, file := range foundFiles {
+			if processedFiles[file] {
+				continue
+			}
+
+			processedNewFiles = true
+			processedFiles[file] = true
+
+			if err := processStackFile(ctx, opts, file); err != nil {
+				return errors.Errorf("Failed to process stack file %s %v", file, err)
+			}
+		}
+
+		if !processedNewFiles {
+			break
+		}
+
+		newFiles, err := listStackFiles(opts, opts.WorkingDir)
+
+		if err != nil {
+			return errors.Errorf("Failed to list stack files %v", err)
+		}
+
+		foundFiles = newFiles
+	}
+
+	return nil
+}
+
+// StackOutput generates the output from the stack files.
+func StackOutput(ctx context.Context, opts *options.TerragruntOptions) (map[string]map[string]cty.Value, error) {
+	opts.Logger.Debugf("Generating output from %s", opts.TerragruntStackConfigPath)
+	opts.TerragruntStackConfigPath = filepath.Join(opts.WorkingDir, defaultStackFile)
+	stackTargetDir := filepath.Join(opts.WorkingDir, stackDir)
+	stackFiles, err := listStackFiles(opts, stackTargetDir)
+
+	if err != nil {
+		return nil, errors.Errorf("Failed to list stack files in %s %v", stackTargetDir, err)
+	}
+
+	unitOutputs := make(map[string]map[string]cty.Value)
+
+	// add default stack file
+	stackFiles = append(stackFiles, opts.TerragruntStackConfigPath)
+
+	for _, path := range stackFiles {
+		stackFile, err := ReadStackConfigFile(ctx, opts, path)
+
+		if err != nil {
+			return nil, errors.New(err)
+		}
+
+		// process each unit and get outputs
+		for _, unit := range stackFile.Units {
+			opts.Logger.Debugf("Processing unit %s", unit.Name)
+
+			dir := filepath.Dir(path)
+			unitDir := filepath.Join(dir, stackDir, unit.Path)
+			output, err := unit.ReadOutputs(ctx, opts, unitDir)
+
+			if err != nil {
+				return nil, errors.New(err)
+			}
+
+			unitOutputs[unit.Name] = output
+		}
+	}
+
+	return unitOutputs, nil
+}
+
+// processStackFile process single stack file.
+func processStackFile(ctx context.Context, opts *options.TerragruntOptions, stackFilePath string) error {
```

Comment:
NIT: We should avoid use consistent language here. "process" doesn't really mean anything in this context. They should all be called "generate", which is what `terragrunt.stack.hcl` files do.